### PR TITLE
Improve and expand TaggableFilter and AtlasStatistics

### PR DESCRIPTION
**Pull request by matthieun**
### Description:

Updated the `TaggableFilter` to use a decision tree structure. Readme available [here](https://github.com/matthieun/atlas/tree/stats/src/main/java/org/openstreetmap/atlas/tags/filters#tag-filtering) ([Permanent link](https://github.com/matthieun/atlas/tree/cc8d841cda29cd1377839d436c967a89cbd604a9/src/main/java/org/openstreetmap/atlas/tags/filters#tag-filtering))

Updating the `AtlasStatistics` class to allow for text-defined coverage metrics, on top of just counts only. This is _not_ an attempt at making the `AtlasStatistics` code better and more maintainable... It is already beyond saving.

### Potential Impact:

More statistics will be printed.

### Unit Test Approach:

Created fake Atlas files using JOSM. Then compare the statistics results to the expected ones.

- waterAtlas.josm.osm:
image
- addressAtlas.josm.osm:
image
- ferryAtlas.josm.osm
image
- refsAtlas.josm.osm
image

### Test Results:

The counts match to the expected counts.

------

In doubt: [Contributing Guidelines](CONTRIBUTING.md)

### CODE REVIEW

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+        return result;
+    }
+
+    private boolean isCounted(final AtlasEntity item, final MultiMap<String, String> andGroup)
+    {
+        for (final String key : andGroup.keySet())
+        {
+            if (item.containsValue(key, andGroup.get(key)))
+            {
+                return true;
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: How about using anyMatch here and below?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+                        {
+                            @Override
+                            protected Iterable<AtlasEntity> getEntities()
+                            {
+                                if (sources.contains("all"))
+                                {
+                                    return getAtlas();
+                                }
+                                final List<Iterable<? extends AtlasEntity>> result = new ArrayList<>();
+                                if (sources.contains("nodes"))
```
Comment (MikeGost) [inline src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: Do you think it's worth pulling these constant out into an enum?

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        {
+            return this.valuesWorkingTogether.toString();
+        }
+    }
+
+    private static final String TYPE_SEPARATOR = ";";
+    private static final String KEYS_SEPARATOR = "|";
+    private static final String VALUES_SEPARATOR = ",";
+    private static final String KEY_VALUE_SEPARATOR = "->";
+    private static final String COUPLED_KEYS_SEPARATOR = "&";
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: Yes the taggable filter re-used this and improved. It should be reused but at that point I am happy letting the stats be independent this way.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    }
+
+    @Override
+    protected CoverageType coverageType()
+    {
+        return this.coverageType;
+    }
+
+    @Override
+    protected Set<String> getKeys(final AtlasEntity item)
```
[the remark this answers is not in the document]
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: It is a mess, I am not sure myself.

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java (hunk 1 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: So all this code comes from CountCoverage that I renamed to SimpleCoverage so I'd rather not touch...

File: src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java (hunk 2 above)
Comment (matthieun) [reply src/main/java/org/openstreetmap/atlas/geography/atlas/statistics/coverage/poi/SimpleCoverage.java]: I can do that..

**Comment by matthieun** [comment]
Closing while I try to figure out more improvements...

File: src/test/java/org/openstreetmap/atlas/tags/filters/TaggableFilterTest.java [hunk 5]
```diff
@@ -57,10 +82,10 @@ public void testParsing()
     @Test
     public void testSerialization() throws ClassNotFoundException
     {
-        final String definition = "amenity->bus_station|highway->BUS_STOP|bus->*^trolleybus->*&public_transport->stop_position,platform,station";
-        final TaggableFilter filter = new TaggableFilter(definition);
+        final String definition = "amenity->bus_station|highway->BUS_STOP|bus->*||trolleybus->*&public_transport->stop_position,platform,station";
```
Comment (jwpgage) [inline src/test/java/org/openstreetmap/atlas/tags/filters/TaggableFilterTest.java]: Looking at the README, backwards compatibility means that the ^ notation should still work, correct? Is there a reason the unit test was changed?

You can tell I don't want to refactor the `AtlasMissingShardVerifier` :)

**Comment by jwpgage** [comment]
Sorry for the late review, it appears my GitHub email notifications still aren't working correctly :/

File: src/test/java/org/openstreetmap/atlas/tags/filters/TaggableFilterTest.java (hunk 5 above)
Comment (matthieun) [reply src/test/java/org/openstreetmap/atlas/tags/filters/TaggableFilterTest.java]: I modified the test to be using the new syntax, however, I added a new test to verify that the backwards compatibility was not dropped: https://github.com/osmlab/atlas/pull/229/files/fd9e83b08b95dd1ca07133fc431f6f2284378765#diff-66d20c24969469313d66a1c180052f0cR20

Comment (jwpgage) [reply src/test/java/org/openstreetmap/atlas/tags/filters/TaggableFilterTest.java]: Thanks, not sure how I missed that!

**Pull request merged**